lib/Conversion/AffineToPipeline/AffineToPipeline.cpp

```diff
@@ -397,48 +431,101 @@ LogicalResult AffineToPipeline::createPipelinePipeline(
 
   // Maintain mappings of values in the loop body and results of stages,
   // initially populated with the iter args.
-  BlockAndValueMapping valueMap;
-  for (size_t i = 0; i < iterArgs.size(); ++i)
-    valueMap.map(forOp.getBody()->getArgument(i),
-                 pipeline.getStagesBlock().getArgument(i));
+  BlockAndValueMapping iterValueMap;
+  for (size_t i = 0; i < forOp.getBody()->getNumArguments(); ++i)
+    iterValueMap.map(forOp.getBody()->getArgument(i),
+                     pipeline.getStagesBlock().getArgument(i));
 
   // Create the stages.
   Block &stagesBlock = pipeline.getStagesBlock();
   builder.setInsertionPointToStart(&stagesBlock);
 
   // Iterate in order of the start times.
   SmallVector<unsigned> startTimes;
-  for (auto group : startGroups)
+  for (const auto &group : startGroups)
     startTimes.push_back(group.first);
   llvm::sort(startTimes);
 
   DominanceInfo dom(getOperation());
+  SmallVector<DenseSet<Value>> registerValues;
+  SmallVector<SmallVector<mlir::Type>> registerTypes;
+  SmallVector<BlockAndValueMapping> valueMaps;
+  DenseMap<Operation *, std::pair<unsigned, unsigned>> pipeTimes;
   for (auto startTime : startTimes) {
     auto group = startGroups[startTime];
-    OpBuilder::InsertionGuard g(builder);
 
     // Collect the return types for this stage. Operations whose results are not
     // used within this stage are returned.
     auto isLoopTerminator = [forOp](Operation *op) {
       return isa<AffineYieldOp>(op) && op->getParentOp() == forOp;
     };
-    SmallVector<Type> stageTypes;
-    DenseSet<Operation *> opsWithReturns;
+
     for (auto *op : group) {
+      if (op->getUsers().empty())
+        continue;
+      unsigned pipeStartTime = -1;
+      unsigned pipeEndTime = 0;
       for (auto *user : op->getUsers()) {
         if (*problem.getStartTime(user) > startTime || isLoopTerminator(user)) {
-          opsWithReturns.insert(op);
-          stageTypes.append(op->getResultTypes().begin(),
-                            op->getResultTypes().end());
+          pipeEndTime = std::max(pipeEndTime, *problem.getStartTime(user));
+          pipeStartTime = std::min(pipeStartTime, *problem.getStartTime(user));
         }
       }
+      // Skip iter if the users are in the same stage
+      if (pipeStartTime > pipeEndTime)
+        continue;
+
+      pipeTimes[op] = std::pair(pipeStartTime, pipeEndTime);
+
+      // Add register stages for each time slice
+      for (unsigned i = registerValues.size(); i <= pipeEndTime; ++i)
+        registerValues.push_back(DenseSet<Value>());
+
+      for (auto result : op->getResults())
+        registerValues.data()[startTime].insert(result);
+
+      // Handling multi-cycle ops here
```

Comment:
I'm not sure about the modeling of this handling. It seems like you intend to have a register in every stage after a multi-cycle op until where it is consumed. Something like this:

```
%1 = stage 0:
  %0 = arith.muli %a, %b
  register %0
%2 = stage 1:
  register %1
%3 = stage 2:
  register %2
stage 3:
  consume(%3)
```

I am not sure what `stage 1` and `stage 2` convey here, if your backend knows `arith.muli` is 3 cycles. It seems like internal implementation details of how the backend implements `arith.muli` have leaked out into the pipeline IR. Why do the internal registers need to be explicit in the IR while the internal combinational logic is still implicit and all represented by the op in `stage 0`?

In other words, if you wanted to register intermediate results for internal stages of multi-cycle ops, I'd expect something like:

```
%1 = stage 0:
  %0 = arith.muli_start %a, %b
  register %0
%3 = stage 1:
  %2 = arith.muli_mid %1
  register %2
%5 = stage 2:
  %4 = arith.muli_end %3
  register %4
stage 3:
  consume(%5)
```

For what it's worth, we made a conscious decision not to explicitly register results of multi-cycle ops, and simply consume them in the correct downstream stage. It is up to the lowering to insert the internal registers to the implementation of the multi-cycle ops. For example, the lowering of muli in Calyx creates internal registers.

Can you explain what you are trying to solve, and how modeling this differently helps?